fastcrypto/src/hash.rs

```diff
@@ -101,52 +102,51 @@ pub trait HashFunction<const DIGEST_LENGTH: usize>: Default {
     /// Compute a single digest from all slices in the iterator in order and consume the hash function.
     fn digest_iterator<K: AsRef<[u8]>, I: Iterator<Item = K>>(iter: I) -> Digest<DIGEST_LENGTH> {
         let mut h = Self::default();
-        iter.into_iter().for_each(|chunk| h.update(chunk.as_ref()));
+        iter.for_each(|item| h.update(item));
         h.finalize()
     }
 }
 
 /// This trait is implemented by all messages that can be hashed.
 pub trait Hash<const DIGEST_LEN: usize> {
     /// The type of the digest when this is hashed.
-    type TypedDigest: Into<Digest<DIGEST_LEN>> + Eq + std::hash::Hash + Copy + fmt::Debug;
+    type TypedDigest: Into<Digest<DIGEST_LEN>> + Eq + std::hash::Hash + Copy + Debug;
 
     fn digest(&self) -> Self::TypedDigest;
 }
 
 /// This wraps a [digest::Digest] as a [HashFunction].
 #[derive(Default)]
-pub struct HashFunctionWrapper<Variant: digest::Digest + 'static, const DIGEST_LEN: usize>(Variant);
+pub struct HashFunctionWrapper<Variant, const DIGEST_LEN: usize>(Variant);
 
 /// This trait allows using a [HashFunctionWrapper] where a [digest::Digest] was expected.
 pub trait ReverseWrapper {
-    type Variant: digest::Digest + 'static + digest::core_api::CoreProxy + OutputSizeUser;
+    type Variant: digest::core_api::CoreProxy + OutputSizeUser;
 }
 
-impl<
-        Variant: digest::Digest + 'static + digest::core_api::CoreProxy + OutputSizeUser,
-        const DIGEST_LEN: usize,
-    > ReverseWrapper for HashFunctionWrapper<Variant, DIGEST_LEN>
+impl<Variant: digest::core_api::CoreProxy + OutputSizeUser, const DIGEST_LEN: usize> ReverseWrapper
+    for HashFunctionWrapper<Variant, DIGEST_LEN>
 {
     type Variant = Variant;
 }
 
-impl<Variant: digest::Digest + 'static + Default, const DIGEST_LEN: usize> HashFunction<DIGEST_LEN>
+impl<Variant: digest::Digest + Default, const DIGEST_LEN: usize> HashFunction<DIGEST_LEN>
     for HashFunctionWrapper<Variant, DIGEST_LEN>
 {
     fn update<Data: AsRef<[u8]>>(&mut self, data: Data) {
         self.0.update(data);
     }
 
     fn finalize(self) -> Digest<DIGEST_LEN> {
-        Digest {
-            digest: self.0.finalize().as_slice().try_into().unwrap(),
-        }
+        let mut digest = [0u8; DIGEST_LEN];
+        self.0
+            .finalize_into(GenericArray::from_mut_slice(&mut digest));
```

Comment:
I think this does exactly what the previous code did, but I think it's a bit clearer like this